tests/test-dladdr.c

```diff
@@ -0,0 +1,177 @@
+ */
+int check_dladdr_by_dlopen( const char *hint, char *libname, char *sym, int should_match )
+{
+    void *library = NULL;
+    void *addr = NULL;
+    int result;
+
+    library = dlopen( libname, RTLD_GLOBAL );
+    if ( library == NULL )
+    {
+        fprintf( stderr, "could not open '%s'\n", libname );
+        return 2;
+    }
+
+    addr = dlsym( library, sym );
+    if ( !addr ) {
+        fprintf( stderr, "could not get address from library '%s' for symbol '%s'\n", libname, sym );
+        return 3;
+    }
+
+    result = check_dladdr( hint, addr, sym, should_match );
+    dlclose( library );
+
+    return result;
+}
+
+#ifdef _WIN32
+#include <windows.h>
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
+/* hide warning "reclared without 'dllimport' attribute" */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wattributes"
+#elif defined(_MSC_VER)
+/* disable warning C4273 inconsistent dll linkage */
+#pragma warning(push)
+#pragma warning(disable: 4273)
+#endif
+/*
+ * Windows API functions decorated with WINBASEAPI are imported from iat.
+ * For testing force linking by import thunk for the following functions
+ */
+HMODULE WINAPI GetModuleHandleA (LPCSTR lpModuleName);
+SIZE_T WINAPI VirtualQuery (LPCVOID lpAddress, PMEMORY_BASIC_INFORMATION lpBuffer, SIZE_T dwLength);
+/* force linking by iat */
+/* WINBASEAPI, which is normally used here, can be overriden by compiler or application, so we cannot use it */
+__declspec(dllimport) HMODULE WINAPI LoadLibraryExA(LPCSTR lpLibFileName, HANDLE hFile, DWORD  dwFlags);
+__declspec(dllimport) int __cdecl atoi(const char *_Str);
+
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
+#pragma GCC diagnostic pop
+#elif defined(_MSC_VER)
+#pragma warning(pop)
+#endif
+#endif
+
+#ifdef _WIN32
+#define PassWithoutSymbolOnWin PassWithoutSymbol
+#define PassWithDifferentAddressOnWin PassWithDifferentAddress
+#else
+#define PassWithoutSymbolOnWin Pass
+#define PassWithDifferentAddressOnWin Pass
+#endif
+
+#if defined(_WIN32) && !defined(DLFCN_WIN32_SHARED)
+#define PassWithSymbolOnSharedBuild PassWithoutSymbol
+#else
+#define PassWithSymbolOnSharedBuild Pass
+#endif
+
+#define UNUSED(x) (void)(x)
+
+int main(int argc, char **argv)
+{
+    UNUSED(argv);
+
+    int  result = 0;
+    if (argc == 2)
+        verbose = 1;
+
+    result = check_dladdr( "null pointer", (void*)0, NULL , NoInfo);
+    result |= check_dladdr( "invalid pointer", (void*)0x125, NULL , NoInfo);
+    result |= check_dladdr( "function from dl library", ((void*)dladdr), "dladdr" , PassWithSymbolOnSharedBuild );
+    result |= check_dladdr( "function from dl library", (void*)dlopen, "dlopen", PassWithSymbolOnSharedBuild );
+    result |= check_dladdr( "function from glibc/msvcrt library", (void*)atoi, "atoi", Pass );
+    result |= check_dladdr( "function from executable", (void*)main, "main", PassWithoutSymbolOnWin );
```

Comment:
Can you add also some test for `static` function to check it has really no symbol?